### PR TITLE
Fix installation path of CMake package files

### PR DESCRIPTION
**Pull request by shaxbee**
Installation path prevented cmake from finding GLFW. It seems that INSTALL_DIRECTORY was correct but path provided to INSTALL command was not updated.
Added GLFW_LIBRARIES export to glfwConfig.cmake.in as well.

Now it is possible to use glfw with externalproject_add and find_package as follows:
```
find_package(glfw3 REQUIRED)
add_executable(foo foo.cpp)
target_link_libraries(foo ${GLFW3_LIBRARY})
```

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -567,7 +567,7 @@ if (GLFW_INSTALL)
 
     install(FILES "${GLFW_BINARY_DIR}/src/glfw3Config.cmake"
                   "${GLFW_BINARY_DIR}/src/glfw3ConfigVersion.cmake"
-            DESTINATION lib${LIB_SUFFIX}/cmake/glfw)
+            DESTINATION ${GLFW_CONFIG_PATH})
 
     install(EXPORT glfwTargets DESTINATION lib${LIB_SUFFIX}/cmake/glfw)
```
Comment (ngvitovitch) [inline CMakeLists.txt]: It may be appropriate to change glfwTargets DESTINATION to ${GLFW_CONFIG_PATH} as well.

File: src/glfw3Config.cmake.in [hunk 2]
```diff
@@ -13,3 +13,4 @@ set_and_check(GLFW3_LIBRARY_DIR "@PACKAGE_CMAKE_INSTALL_PREFIX@/lib@LIB_SUFFIX@"
 
 find_library(GLFW3_LIBRARY "@GLFW_LIB_NAME@" HINTS ${GLFW3_LIBRARY_DIR})
 
+set(GLFW_LIBRARIES "@GLFW_LIBRARIES@")
```
Comment (ngvitovitch) [inline src/glfw3Config.cmake.in]: While we're pluralizing things, set GLFW_INCLUDE_DIRS and GLFW_LIBRARY_DIRS as well.

**Comment by shaxbee** [comment]
Fixed the glfwTargets
Where should I get the INCLUDE_DIRS / LIBRARY_DIRS from?

**Comment by ngvitovitch** [comment]
I would imaging setting them to ${GLFW3_INCLUDE_DIR} and ${GLFW3_LIBRARY_DIR}, to maintain SSOT.

But before implementing this change, remind me of the rational of defining two sets of _LIBRARY, _INCLUDE_DIR and _LIBRARY_DIR, one with glfw and the other glfw3?  Is this to maintain some backwards compatibility?

**Comment by shaxbee** [comment]
No, i just wanted to keep it uniform with package name. What is the reason config prefix is glfw3 not simply glfw?

**Comment by ngvitovitch** [comment]
I believe this was done intentionally, as not to cause confusion with installations of the now-legacy glfw2.x.

**Comment by ngvitovitch** [comment]
A quick build of your master branch yields the following:
```
C:\Users\nvitovitch\Projects\glfw\build\glfw-shaxbee>cmake -G"Visual Studio 10 Win64" -DCMAKE_PREFIX_PATH=C:\Users\nvitovitch\Projects\glfw\pkg-shaxbee -DCMAKE_INSTALL_PREFIX=C:\Users\nvitovitch\Projects\glfw\pkg-shaxbee ..\..\code\glfw-shaxbee
-- The C compiler identification is MSVC 16.0.40219.1
-- Check for working C compiler using: Visual Studio 10 Win64
-- Check for working C compiler using: Visual Studio 10 Win64 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Found OpenGL: opengl32
-- Looking for include file pthread.h
-- Looking for include file pthread.h - not found
-- Found Threads: TRUE
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
-- Using Win32 for window creation
-- Using WGL for context creation
-- Configuring done
CMake Error: install(EXPORT "glfwTargets") given absolute DESTINATION "C:/Users/nvitovitch/Projects/glfw/pkg-shaxbee" but the export references an installation of target "glfw" which has relative DESTINATION "lib".
CMake Error: install(EXPORT "glfwTargets") given absolute DESTINATION "C:/Users/nvitovitch/Projects/glfw/pkg-shaxbee" but the export references an installation of target "glfw" which has relative DESTINATION "lib".
CMake Error: install(EXPORT "glfwTargets") given absolute DESTINATION "C:/Users/nvitovitch/Projects/glfw/pkg-shaxbee" but the export references an installation of target "glfw" which has relative DESTINATION "lib".
CMake Error: install(EXPORT "glfwTargets") given absolute DESTINATION "C:/Users/nvitovitch/Projects/glfw/pkg-shaxbee" but the export references an installation of target "glfw" which has relative DESTINATION "lib".
-- Generating done
-- Build files have been written to: C:/Users/nvitovitch/Projects/glfw/build/glfw-shaxbee
```
I think the correct fix is to change GLFW_CONFIG_PATH to a relative path, namely `set(GLFW_CONFIG_PATH "lib${LIB_SUFFIX}/cmake/glfw")`.

To address the original intent of your pull request, linking external projects with ease, `target_link_libraries(glfw LINK_PUBLIC ${glfw_LIBRARIES})` might exhibit the desired behaviour.  I'm going to run some tests and get back to you :)

**Comment by ngvitovitch** [comment]
I spent some significant time reviewing cmake, cmake targets, etc. and I have a solution.  I will try an submit a pull request to your master branch by 5EST!  Afterwards we should do some prelim builds and tests, just to be sure everything is working.

**Comment by shaxbee** [comment]
Merged @ngvitovitch changes into pull request - they add link libraries as transitive dependency, I've added include directory as well. This change removes need to set include_directories and target_link_libraries (for transitive dependencies).

**Comment by shaxbee** [comment]
Tested on Mac OS X.

**Comment by ngvitovitch** [comment]
Will test on Windows, Linux at my earliest convenience.  Could you confirm that the generated glfw3Targets* files contain no absolute paths?  I am mildly suspicious of the latest target_include_dirs().

**Comment by shaxbee** [comment]
It does contain absolute path because relative path nor source/binary
prefixed path is not allowed in there:
```
  Target "glfw" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "/Users/shaxbee/hasiok/glfw-cmake-package/src/../include"

  which is prefixed in the source directory.
```

**Comment by ngvitovitch** [comment]
Not true, the absolute path is not necessary and moreover it breaks portability of the package.  Example:
1. Install glfw to \<pkg_dir\>,
2. Relocate \<pkg_dir\>
3. Configure and build \<my_app\> using CMake - building fails due to missing headers b/c the path baked into glfw3Targets.cmake is now invalid.

Check out this example using CMake generator expressions: http://www.cmake.org/cmake/help/v3.0/command/target_include_directories.html

Also, target_include_directories was introduced in CMake 2.8.11 and will cause an error when building with glfw's minimum required version of CMake (2.8.9).  If this command is included, it needs to be if-guarded for older versions of CMake.

**Comment by shaxbee** [comment]
@elmindreda @ngvitovitch: BTW. Since C++11 is already enforced, and Ubuntu 14.04 LTS comes with CMake 2.8.12, maybe it would worth to bump cmake version to provide uniform build.

**Comment by ngvitovitch** [comment]
@elmindreda requesting feedback on PR

**Comment by elmindreda** [comment]
Will do after 3.1.1. Only one issue remaining.

**Comment by ngvitovitch** [comment]
@elmindreda huzzah for the 3.1.1 release!

**Comment by elmindreda** [comment]
Please make your commit messages [well formed](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). Subject line max 50 characters, remaining lines max 72 characters, empty lines between paragraphs.

**Comment by shaxbee** [comment]
@elmindreda Reformatted commit messages

**Comment by shaxbee** [comment]
@elmindreda PR ready for merge

**Comment by elmindreda** [comment]
@shaxbee I know. I'm just catching up.

**Comment by elmindreda** [comment]
I'll be pushing a lot of CMake-related fixes in the coming week that'll probably make this flag as unmergeable but there's no need to do anything because it's just side-effects of preparing to merge parts of your work.

**Comment by shaxbee** [comment]
Give me a shout when you'll do that - I'll rebase PR.

**Comment by larshg** [comment]
@shaxbee I cannot get this to work on my Win7 using Cmake generating solutions for VS 2013. The 
variable ```${GLFW3_LIBRARY}``` is just empty. Do you set any variable like ```GLFW_FOUND``` - some of the Cmake scripts is beyond my knowledge and seem quite complicated :)

I also had to add ```${CMAKE_BINARY_DIR}``` infront of the generated config files, since I build out of source folder.

```
install(
        FILES 
            ${GLFW_INSTALL_CONFIG_DIR}/glfw3Config.cmake
            glfw3ConfigVersion.cmake
            ${CMAKE_BINARY_DIR}/glfw3Config.cmake
            ${CMAKE_BINARY_DIR}/glfw3ConfigVersion.cmake
        DESTINATION ${GLFW_INSTALL_CONFIG_DIR})
```

**Comment by shaxbee** [comment]
@larshg Could you be so kind and try it now?

**Comment by larshg** [comment]
Works like a charm, but it would be nice with some cmake output of GLFW3 actually being found.

**Comment by shaxbee** [comment]
@larshg: CMake provides standard message when find_package is being used.

**Comment by tamaskenez** [comment]
@elmindreda , as I understand you're going to cherry-pick parts of this PR. I've been using @shaxbee's master for a while, let me comment on a specific issue with it.

Shaxbee uses CMake's `export(PACKAGE)` feature to register the build tree to the CMake user package registry. This feature is just like any other global automagic: it's very convenient if it does what you expect and source of hard to find bugs it does not. In our systems (local and CI nodes) we keep on having issues when due to a misconfiguration certain projects tend to find completely unrelated glfw installations (instead of not finding glfw or giving an error message).

So I'd vote against this feature, or alternatively, make it optional by adding this line in the `CMakeLists.txt`:

    option(CMAKE_EXPORT_NO_PACKAGE_REGISTRY "" ON)

so it can be enabled explicitly at configuration time for those who needs it with `cmake ... -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=OFF ...`

**Comment by shaxbee** [comment]
Tamas: I have mixed feelings about package export too and have no problem
removing it from PR as it doesn't seem to be standard practice.

**Comment by shaxbee** [comment]
@tamaskenez removed package export, added debug postfix change.

**Comment by shaxbee** [comment]
Rebased to upstream

**Comment by shaxbee** [comment]
Rebased, added Travis CI to track CMake <3.0 compatibility

**Comment by shaxbee** [comment]
@elmindreda Any ETA on merging this?

**Comment by elmindreda** [comment]
I've picked out bits I wanted from this patch set. There's a lot of good stuff here, but also some bugs, things I won't be merging and things I'd already fixed while learning modern CMake so I could tackle this issue.

I've tried to retain proper credits for all the pieces, but not at the expense of clear patches without backtracking. Thank you so much, both of you, and sorry for taking so long to get to this.

**Comment by elmindreda** [comment]
Also merged bits of the Travis and Appveyor CI files. Thanks!

**Pull request closed without merging**